napps/kytos/of_core/main.py

```diff
@@ -107,31 +109,100 @@ def handle_features_reply(self, event):
+        data_len = len(data)
+        pkts = []
+        while data_len > 3:
+            length_field = struct.unpack('!H', data[2:4])[0]
+            if data_len >= length_field:
+                pkts.append(data[:length_field])
+                data = data[length_field:]
+                data_len = len(data)
+        return pkts, data
+
+    @listen_to('kytos/core.server.data.in')
+    def handle_new_server_data(self, event):
         """Handle a RawEvent and generate a kytos/core.messages.in.* event.
 
         Args:
             event (KytosEvent): RawEvent with openflow message to be unpacked
         """
-        log.debug('RawOpenFlowMessage received by RawOFMessage handler')
-
-        # creates an empty OpenFlow Message based on the message_type defined
-        # on the unpacked header.
-        message = new_message_from_header(event.content['header'])
-        binary_data = event.content['binary_data']
-
-        # The unpack will happen only to those messages with body beyond header
-        if binary_data and len(binary_data) > 0:
-            message.unpack(binary_data)
-        log.debug('RawOpenFlowMessage unpacked')
+        packets, event.source.remaining_data = \
+            self.of_slicer(event.source.remaining_data +
+                           event.content['new_data'])
+
+        # if it is a new connection, we are expecting a hello packet
+        if event.source.state == event.source.STATES.NEW:
+            try:
+                hello_packet = packets.pop(0)
+            except IndexError:
+                pass
+            else:
+                if self.validate_hello(hello_packet):
+                    event.source.state = event.source.STATES.SETUP
+                    event.source.protocol.state = 'hello'
+
+                    hello_message = self.unpack(hello_packet)
+                    self._handle_hello(hello_message, event.source)
+                else:
+                    log.debug('Invalid OF Hello message')
+                    event.source.protocol.state = 'hello_invalid'
+                    event.source.close()
+                    return
+
+        for packet in packets:
+            log.debug('New Raw Openflow Message')
+            log.debug(packet)
+            message = self.unpack(packet)
+            log.debug('RawOpenFlowMessage unpacked')
+            log.debug(message)
+            self.emmit_message_in(message, event.source)
+
+    def unpack(self, packet):
+        # creates an empty OpenFlow Message based on the message_type
+        # defined on the unpacked header.
+        try:
+            header = Header()
+            header.unpack(packet[:8])
+            message = new_message_from_header(header)
+            binary_data = packet[8:]
+            if binary_data:
+                message.unpack(binary_data)
+        except Exception as e:
+            log.info('Could not unpack message', packet)
+            raise e
+        else:
+            # TODO: validate message
```

Comment:
Instead of adding a `TODO` It is preferable to create an new Issue indication the task that need to be done and eventually linking the method/function/line.